core/src/main/java/org/infinispan/container/versioning/irac/IracEntryVersion.java

```diff
@@ -0,0 +1,307 @@
+import org.infinispan.protostream.descriptors.Type;
+
+/**
+ * An entry version for the IRAC algorithm (async cross site replication).
+ * <p>
+ * It is represented as a vector clock where each site keeps it version.
+ * <p>
+ * The site version is composed as a pair (topology id, version).
+ *
+ * @author Pedro Ruivo
+ * @see TopologyIracVersion
+ * @since 11.0
+ */
+@ProtoTypeId(ProtoStreamTypeIds.IRAC_VERSION)
+public class IracEntryVersion {
+
+   private final Map<String, TopologyIracVersion> vectorClock;
+
+   public IracEntryVersion(Map<String, TopologyIracVersion> vectorClock) {
+      this.vectorClock = Objects.requireNonNull(vectorClock);
+   }
+
+   @ProtoFactory
+   static IracEntryVersion protostreamFactory(List<MapEntry> entries) {
+      Map<String, TopologyIracVersion> vectorClock = entries.stream()
+            .collect(Collectors.toMap(mapEntry -> mapEntry.site, mapEntry -> mapEntry.version));
+      return new IracEntryVersion(vectorClock);
+   }
+
+   /**
+    * Converts this instance to a {@link Map}.
+    * <p>
+    * The map cannot be modified!.
+    *
+    * @return The {@link Map} representation of this version.
+    */
+   public Map<String, TopologyIracVersion> toMap() {
+      return Collections.unmodifiableMap(vectorClock);
+   }
+
+   /**
+    * Iterates over all entries of this version as pairs (site name, site version).
+    *
+    * @param consumer The {@link BiConsumer}.
+    */
+   public void forEach(BiConsumer<String, TopologyIracVersion> consumer) {
+      vectorClock.forEach(consumer);
+   }
+
+   /**
+    * Compares this instance with another {@link IracEntryVersion} instance.
+    * @param other The other {@link IracEntryVersion} instance.
+    * @return A {@link InequalVersionComparisonResult} instance with the compare result.
+    */
+   public InequalVersionComparisonResult compareTo(IracEntryVersion other) {
+      VectorClock vectorClock = new VectorClock();
+      this.forEach(vectorClock::setOurs);
+      other.forEach(vectorClock::setTheirs);
+
+      Merger merger = Merger.NONE;
+      for (VersionCompare v : vectorClock.values()) {
+         merger = merger.accept(v);
+      }
+      return merger.result();
+   }
+
+   @Override
+   public String toString() {
+      List<String> entries = new LinkedList<>();
+      vectorClock.forEach((site, version) -> entries.add(site + "=" + version));
+      return "(" + String.join(", ", entries) + ")";
+   }
+
+   @Override
+   public boolean equals(Object o) {
+      if (this == o) {
+         return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+         return false;
+      }
+
+      IracEntryVersion version = (IracEntryVersion) o;
+
+      return vectorClock.equals(version.vectorClock);
+   }
+
+   @Override
+   public int hashCode() {
+      return vectorClock.hashCode();
+   }
+
+   @ProtoField(number = 1, collectionImplementation = ArrayList.class)
+   List<MapEntry> entries() {
```

Comment:
If we place this below the `@ProtoFactory` method it makes the marshalling logic easier to read as it's all in one place.

Reply:
:+1: